middleware/metrics.go

```diff
@@ -0,0 +1,71 @@
+package middleware
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/gorilla/mux"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type responseWriter struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+func NewResponseWriter(w http.ResponseWriter) *responseWriter {
+	return &responseWriter{w, http.StatusOK}
+}
+
+func (rw *responseWriter) WriteHeader(code int) {
+	rw.statusCode = code
+	rw.ResponseWriter.WriteHeader(code)
+}
+
+var totalRequests = prometheus.NewCounterVec(
+	prometheus.CounterOpts{
+		Name: "http_requests_total",
+		Help: "Number of get requests.",
+	},
+	[]string{"path"},
+)
+
+var responseStatus = prometheus.NewCounterVec(
+	prometheus.CounterOpts{
+		Name: "response_status",
+		Help: "Status of HTTP response",
+	},
+	[]string{"status"},
+)
+
+var httpDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+	Name: "http_response_time_seconds",
+	Help: "Duration of HTTP requests.",
+}, []string{"path"})
+
+func RegisterPrometheus() {
+	prometheus.Register(totalRequests)
+	prometheus.Register(responseStatus)
+	prometheus.Register(httpDuration)
+}
+
+// NewPrometheusMiddleware creates a middleware which produces metrics about requests.
+// Should be attached to a mux or other router.
+func NewPrometheusMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		route := mux.CurrentRoute(r)
+		path, _ := route.GetPathTemplate()
+
+		timer := prometheus.NewTimer(httpDuration.WithLabelValues(path))
+		rw := NewResponseWriter(w)
```

Comment:
Do we need to wrap this? Can we not read rw.Header() after write?

Reply:
I just checked this and it seems like this is one reason I see people using the `negroni` package, to solve this exact issue haha. Can we use that instead?

Reply:
For sure, I'm not super familiar with the negroni package (it looks really good though), do you have an example of what you're thinking here I can reference?

Reply:
https://stackoverflow.com/a/63896952
^This is an example of exactly the change I think haha.

Reply:
I think I did what was being looked for here. The route/path extraction with mux is used because we need to get them for prometheus labels, and negroni call replaces the call I had here, and removes 3 blocks of boilerplate I had to write to enable it, so that's a good save.